mock: add README to tracing-mock (#2362)
tokio-rs/tracing: tracing-mock/README.md

![Tracing — Structured, application-level diagnostics][splash]

[splash]: https://raw.githubusercontent.com/tokio-rs/tracing/master/assets/splash.svg

# tracing-mock

Utilities for testing [`tracing`][tracing] and crates that uses it.

[![Documentation (master)][docs-master-badge]][docs-master-url]
[![MIT licensed][mit-badge]][mit-url]
[![Build Status][actions-badge]][actions-url]
[![Discord chat][discord-badge]][discord-url]

[Documentation][docs-master-url] | [Chat][discord-url]

[docs-master-badge]: https://img.shields.io/badge/docs-master-blue
[docs-master-url]: https://tracing-rs.netlify.com/tracing_mock
[mit-badge]: https://img.shields.io/badge/license-MIT-blue.svg
[mit-url]: LICENSE
[actions-badge]: https://github.com/tokio-rs/tracing/workflows/CI/badge.svg
[actions-url]:https://github.com/tokio-rs/tracing/actions?query=workflow%3ACI
[discord-badge]: https://img.shields.io/discord/500028886025895936?logo=discord&label=discord&logoColor=white
[discord-url]: https://discord.gg/EeF3cQw

## Overview

> [`tracing`] is a framework for instrumenting Rust programs to collect
> structured, event-based diagnostic information. This crate provides
> constructors for mock `tracing` objects to allow testing of `tracing`
> itself as well as crates that use it.

Reviewer: ```suggestion
[`tracing`] is a framework for instrumenting Rust programs to collect
structured, event-based diagnostic information. `tracing-mock` provides
mock `tracing` objects that are useful in testing `tracing`  itself 
and crates that use `tracing`.
```

*Compiler support: [requires `rustc` 1.49+][msrv]*

[msrv]: #supported-rust-versions

## Usage

> The `tracing-mock` crate provides a mock `subscriber` which can
> verify the spans and events sent. It is inteded for use in tests
> to ensure that your crate is tracing at the corrrect time with the
> correct data.

Reviewer: ```suggestion
`tracing-mock` crate provides a mock `Subscriber` that allows asserting on 
the order and contents of spans and events.
```

As `tracing-mock` isn't available on [crates.io](https://crates.io/)
yet, you must import it via git. It is important that you also override
the source of any `tracing` crates that are transient dependencies. For
example, the `Cargo.toml` for your test crate could contain:

```toml
[dependencies]
lib-under-test = 1.0 # depends on `tracing`

[dev-dependencies]
tracing-mock = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x", version = "0.1" }
tracing = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x", version = "0.1" }

[patch.crates-io]

> tracing = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x" }

Reviewer: I think we'll need to also override tracing-core, but please confirm!

Author: Yes, you're right! It's not always necessary, but if `lib-under-test` depends on `tracingc-core` directly, then it is - so it should be included here.

```

> 

Reviewer: ```suggestion
## Examples

```

Below is a simple example of checking for an event with only a

> message.

Reviewer: Below is an example that checks that an event contains a message:

```rust
use tracing::subscriber::with_default;
use tracing_mock::{event, field, subscriber};

fn yak_shaving() {
    tracing::info!("preparing to shave yaks");
}

#[test]
fn traced_event() {
    let (subscriber, handle) = subscriber::mock()
        .event(event::mock().with_fields(field::msg("preparing to shave yaks")))
        .done()
        .run_with_handle();

    with_default(subscriber, || {
        yak_shaving();
    });

    handle.assert_finished();
}
```

> The following is a more complex example taking the full yak shaving example
> from the `tracing` README and additionally instrumenting the called function
> with a span.

Reviewer: ```suggestion
Below is a slightly more complex example. `tracing-mock` asserts that, in order:
- a span is created with a single field/value pair and entered
- an event is created with the field `number_of_yaks`, a corresponding value of 3, and the message "preparing to shave yaks", and nothing else.
- an event is created with the field `all_yaks_shaved`, a corresponding value of `true`, and the message "yak shaving completed".
- the span was exited.
```

note, i think it'd be a good idea to assert that the span also _closed_.

Author: By closed, do you mean `.drop_span()`?

#[test]
fn yak_shaving_traced() {
    let yak_count: u32 = 3;
    let span = span::mock().named("yak_shaving");

    let (subscriber, handle) = subscriber::mock()
        .new_span(
            span.clone()
                .with_field(field::mock("number_of_yaks").with_value(&yak_count).only()),
        )
        .enter(span.clone())
        .event(
            event::mock().with_fields(
                field::mock("number_of_yaks")
                    .with_value(&yak_count)
                    .and(field::msg("preparing to shave yaks"))
                    .only(),
            ),
        )
        .event(
            event::mock().with_fields(
                field::mock("all_yaks_shaved")
                    .with_value(&true)
                    .and(field::msg("yak shaving completed."))
                    .only(),
            ),
        )
        .exit(span.clone())
        .done()
        .run_with_handle();

    with_default(subscriber, || {
        yak_shaving(yak_count);
    });

    handle.assert_finished();
}
```

The full code for both examples can be found in the [readme.rs](tests/readme.rs)

> tests.

Reviewer: I don't linking to this is strictly necessary.

![Tracing — Structured, application-level diagnostics][splash]

[splash]: https://raw.githubusercontent.com/tokio-rs/tracing/master/assets/splash.svg

# tracing-mock

Utilities for testing [`tracing`][tracing] and crates that uses it.

[![Documentation (master)][docs-master-badge]][docs-master-url]
[![MIT licensed][mit-badge]][mit-url]
[![Build Status][actions-badge]][actions-url]
[![Discord chat][discord-badge]][discord-url]

[Documentation][docs-master-url] | [Chat][discord-url]

[docs-master-badge]: https://img.shields.io/badge/docs-master-blue
[docs-master-url]: https://tracing-rs.netlify.com/tracing_mock
[mit-badge]: https://img.shields.io/badge/license-MIT-blue.svg

> [mit-url]: LICENSE

Reviewer 2: I think we need to change this to an absolute link to the file in the GitHub repo, since this link breaks when included in the RustDoc: https://app.netlify.com/sites/tracing-rs/deploys/636972d48850b90009f66194#L241-L251

[actions-badge]: https://github.com/tokio-rs/tracing/workflows/CI/badge.svg
[actions-url]:https://github.com/tokio-rs/tracing/actions?query=workflow%3ACI
[discord-badge]: https://img.shields.io/discord/500028886025895936?logo=discord&label=discord&logoColor=white
[discord-url]: https://discord.gg/EeF3cQw

## Overview

[`tracing`] is a framework for instrumenting Rust programs to collect

> structured, event-based diagnostic information. `tracing-mock` provides
> mock `tracing` objects that are useful in testing `tracing`  itself 
> and crates that use `tracing`.

Reviewer 2: nit, take it or leave it: I might consider rewording this a bit:
```suggestion
structured, event-based diagnostic information. `tracing-mock` provides
tools for making assertions about what `tracing` diagnostics are emitted
by code under test.
```
or something? not a blocker!

*Compiler support: [requires `rustc` 1.49+][msrv]*

[msrv]: #supported-rust-versions

## Usage

`tracing-mock` crate provides a mock `Subscriber` that allows asserting on 

> the order and contents of spans and events.

Reviewer 2: nit, take it or leave it: _mayyyyybe_ "`Subscriber`", "spans", and "events" could link to the `tracing` docs for those types? not a big deal though.

As `tracing-mock` isn't available on [crates.io](https://crates.io/)
yet, you must import it via git. It is important that you also override
the source of any `tracing` crates that are transient dependencies. For
example, the `Cargo.toml` for your test crate could contain:

```toml
[dependencies]

> lib-under-test = 1.0 # depends on `tracing`

Reviewer 2: nit: the crate's version needs to be a string:
```suggestion
lib-under-test = "1.0" # depends on `tracing`
```

[dev-dependencies]
tracing-mock = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x", version = "0.1" }
tracing = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x", version = "0.1" }

[patch.crates-io]
tracing = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x" }
tracing-core = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x" }

> ```

Reviewer 2: since the intention is to release the crate in the near future, i would probably not have written all this when it'll be  deleted when the crate is published, but...we may as well keep it, now that you've added it!

Author: Yeah, I was perhaps being a little pesimistic here and providing help for anyone wanting to use `tracing-mock` now, getting the docs written is going to take a bit more time and this would have helped me when I first started looking at using `tracing-mock`

    tracing::info!(
        all_yaks_shaved = number_shaved == number_of_yaks,
        "yak shaving completed."
    );
}

#[test]
fn yak_shaving_traced() {
    let yak_count: u32 = 3;
    let span = span::mock().named("yak_shaving");

    let (collector, handle) = collector::mock()
        .new_span(
            span.clone()
                .with_field(field::mock("number_of_yaks").with_value(&yak_count).only()),
        )
        .enter(span.clone())
        .event(
            event::mock().with_fields(
                field::mock("number_of_yaks")
                    .with_value(&yak_count)
                    .and(field::msg("preparing to shave yaks"))
                    .only(),
            ),
        )
        .event(
            event::mock().with_fields(
                field::mock("all_yaks_shaved")
                    .with_value(&true)
                    .and(field::msg("yak shaving completed."))
                    .only(),
            ),
        )
        .exit(span.clone())
        .close_span(span)
        .done()
        .run_with_handle();

    with_default(collector, || {
        yak_shaving(yak_count);
    });

    handle.assert_finished();
}
```

> The full code for both examples can be found in the readme.rs tests.

Reviewer 2: since these are now being run as doctests, this is no longer accurate, right?

Author: Correct, removed.

[![Discord chat][discord-badge]][discord-url]

[Documentation][docs-master-url] | [Chat][discord-url]

[docs-master-badge]: https://img.shields.io/badge/docs-master-blue
[docs-master-url]: https://tracing-rs.netlify.com/tracing_mock
[mit-badge]: https://img.shields.io/badge/license-MIT-blue.svg
[mit-url]: https://github.com/tokio-rs/tracing/blob/master/tracing-mock/LICENSE
[actions-badge]: https://github.com/tokio-rs/tracing/workflows/CI/badge.svg
[actions-url]:https://github.com/tokio-rs/tracing/actions?query=workflow%3ACI
[discord-badge]: https://img.shields.io/discord/500028886025895936?logo=discord&label=discord&logoColor=white
[discord-url]: https://discord.gg/EeF3cQw

## Overview

[`tracing`] is a framework for instrumenting Rust programs to collect
structured, event-based diagnostic information. `tracing-mock` provides
tools for making assertions about what `tracing` diagnostics are emitted
by code under test.

*Compiler support: [requires `rustc` 1.49+][msrv]*

[msrv]: #supported-rust-versions

## Usage

`tracing-mock` crate provides a mock
[`Collector`](https://tracing-rs.netlify.app/tracing/#collectors)
that allows asserting on the order and contents of

> [spans]https://tracing-rs.netlify.app/tracing/#spans) and

Reviewer 2: this should fix the documentation build failure:
```suggestion
[spans](https://tracing-rs.netlify.app/tracing/#spans) and
```

> [events](https://tracing-rs.netlify.app/tracing/#events).

Reviewer 2: rather than linking to the netlify `master` docs here, i think we should probably be linking to the release docs on docs.rs, at least in the release version of these docs...we can change them to point to docs.rs when we do the v0.1.x backport, if you think having this version of the documentation point at the `v0.2.x` documentation for `tracing`, but it's a thing we'll need to remember...

Author: My idea was to link to both the docs.rs docs and the master docs once we publish on crates.io, as that's what is done in the other crates in this repo. Since we don't have the docs.rs ones yet, I left it out entirely.
